### PR TITLE
MRG, ENH: Add autobad detection like MF

### PR DESCRIPTION
**Pull request by larsoner**
At the sprint @olafhauk mentioned that not having `-autobad` support was what he was missing in his Python pipeline. This adds support for it.

### CODE REVIEW

**Comment by larsoner** [comment]
https://16211-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/preprocessing/plot_60_maxwell_filtering_sss.html

**Comment by drammock** [comment]
tutorial LGTM.  I didn't really have time for more than a skim of the code changes.

**Comment by agramfort** [comment]
@larsoner did you test on a number of datasets to which point this implementation matches with maxfilter output?

**Comment by agramfort** [comment]
@olafhauk maybe you can give it a try?

**Comment by agramfort** [comment]
I can test on cam-can data but I would need the bads found by original maxfilter program. @olafhauk @dengemann @SherazKhan do you know where I can find this?

File: mne/preprocessing/maxwell.py [hunk 1]
```diff
@@ -534,18 +555,27 @@ def maxwell_filter(raw, origin='auto', int_order=8, ext_order=3,
...
                 mm_in = np.dot(pS_decomp[:n_use_in],
                                orig_data[:, rel_start:rel_stop])
 
                 # Our output data
                 if not st_only:
+                    if reconstruct == 'in':
+                        proj = S_recon.take(reg_moments[:n_use_in], axis=1)
+                        mult = mm_in
+                    else:
+                        assert reconstruct == 'orig'
```
Comment (wmvanvliet) [inline mne/preprocessing/maxwell.py]: This is an interesting construct I haven't seen before. Why not `elif reconstruct == 'orig'` and have an `else` at the bottom?

File: mne/preprocessing/maxwell.py [hunk 2]
```diff
@@ -1851,3 +1883,104 @@ def _trans_sss_basis(exp, all_coils, trans=None, coil_scale=100.):
...
+    should:
+
+    1. Band-pass filter from 0.1 to 50 Hz, and optionally decimate.
+       Because only peaks matter, you can for example :meth:`~mne.io.Raw.copy`
+       your raw instance, :func:`~mne.io.Raw.resample` to 100 Hz, and then
+       :func:`~mne.io.Raw.filter` to high-pass at 0.1 Hz.
+    2. Run :func:`mne.preprocessing.mark_flat` to exclude bad channels
+       (and mark flat ones).
+
+    .. versionadded:: 0.20
```
Comment (wmvanvliet) [inline mne/preprocessing/maxwell.py]:
```
See also
--------
mark_flat
maxwell_filter
```

**Comment by agramfort** [comment]
@wmvanvliet if you have access to original maxfilter code can you test this PR on some of your dataset? thanks

**Comment by wmvanvliet** [comment]
What procedure do you want me to use to check maxwell vs mne output? Here is my first attempt, giving me different bad channels: https://gist.github.com/wmvanvliet/a3944ad425f28e33fba836315f78169d

**Comment by agramfort** [comment]
yes I would like a public quantified evaluation of the differences so we
know where we stand

>

**Comment by wmvanvliet** [comment]
Here are some results: https://pastebin.com/KafKgLH3
Output is quite different.

**Comment by larsoner** [comment]
For flats you need to use `mne.preprocessing.mark_flat` as a first step. Without it the reconstructions will be different so differences are expected. I can add this to the instructions. BTW it's much faster if you do `raw.resample(100).filter(0.1, None)` rather than just `raw.filter(0.1, 50)`.

**Comment by larsoner** [comment]
@wmvanvliet you might also find that changing the `duration` and/or `limit` arguments in MNE will get you closer to what MaxFilter produces. When I ran things it seemed like sometimes it was using a `duration=5` type of processing, and the deviation values it found were always larger than the ones I got in MNE, so maybe something like `duration=5, limit=5` would get you closer.

**Comment by wmvanvliet** [comment]
With `limit=5`, MNE was detecting too many bads compared to maxfilter. So I tried `duration=5, limit=7`. Results are here: https://pastebin.com/KV9Yzgjn. Code for the script is here: https://gist.github.com/wmvanvliet/a3944ad425f28e33fba836315f78169d

**Comment by agramfort** [comment]
can you cook a summary figure of this? fraction of time where they agree?
summary of average overlaps between sets?

>

**Comment by wmvanvliet** [comment]
How's this? https://pastebin.com/vcKwRbSf

**Comment by wmvanvliet** [comment]
Powerline is at 50Hz, yes

**Comment by larsoner** [comment]
Also did you make sure the params were the same like `origin`, `cross_talk`, `fine_calibration`, etc.? I doubt it would make a big difference but it might make some difference

**Comment by wmvanvliet** [comment]
See the script: https://gist.github.com/wmvanvliet/a3944ad425f28e33fba836315f78169d

**Comment by larsoner** [comment]
- You should probably use `origin=(0., 0., 0.04)` in MNE (assuming these are subject files and not empty room) -- we use a dig fit by default and they use a fixed pos
- Either pass `fine_calibration` and `cross_talk` in MNE or (easier) pass `-ctc off -cal off` in the MF call -- by default MF for a given site will use built-in cross-talk and fine cal
- If you really want to reduce differences in the expansions, use `regularize=None, bad_condition='ignore'` in MNE and `-regularize off` in MF because our regularization behaves slightly differently
- If you do decide to re-run, I'd also do `resample(98)` just in case some line noise is sticking around right at Nyquist, since I'm not sure how MF actually does its lowpass.

**Comment by olafhauk** [comment]
> 
> 
> @olafhauk maybe you can give it a try?

I've just come back from another trip. You want me to test the autobad option in maxfilter?

**Comment by olafhauk** [comment]
> 
> 
> I can test on cam-can data but I would need the bads found by original maxfilter program. @olafhauk @dengemann @SherazKhan do you know where I can find this?
Do you already have an answer to this? They have just appointed a new CamCan administrator who may be able to provide this info.

**Comment by wmvanvliet** [comment]
That improved things a little: https://pastebin.com/M1tdsz6f

**Comment by larsoner** [comment]
> That improved things a little: 

80 hits, 14 misses, 46 false alarms (and 9958 correct rejections) -- getting better at least! I emailed Jukka to see if he has some insight into where other potential differences may arise, but it might take a bit for him to get back to me. I'll mark this WIP in the meantime

> Do you already have an answer to this? They have just appointed a new CamCan administrator who may be able to provide this info.

@olafhauk no we do not have this info yet

**Comment by olafhauk** [comment]
> 
> 
> > That improved things a little:
> 
> 80 hits, 14 misses, 46 false alarms (and 9958 correct rejections) -- getting better at least! I emailed Jukka to see if he has some insight into where other potential differences may arise, but it might take a bit for him to get back to me. I'll mark this WIP in the meantime
> 
> > Do you already have an answer to this? They have just appointed a new CamCan administrator who may be able to provide this info.
> 
> @olafhauk no we do not have this info yet

If you have already applied for access to CamCan MEG data via the CamCan web-site, you should be able to find the log-files with bad channel information on your temporary CBU account. Otherwise you need to apply for access to these data. I don't work with CamCan data myself at the moment I'm afraid. For queries it's probably best for the person who applied for CamCan access to write to rik.henson@mrc-cbu.cam.ac.uk directly.

**Comment by larsoner** [comment]
@wmvanvliet I don't have any lines that look like `Detected 2 flat channels` what MaxFilter version are you using? I have `Revision: 2.2.15`

**Comment by wmvanvliet** [comment]
I have $Revision: 2.2.15 Neuromag maxfilter Dec 11 2012 14:48:36 $

**Comment by larsoner** [comment]
So it turns out that MaxFilter does some stuff that we can and probably should avoid. The reading, filtering, and downsampling is tied to the number of tags and number of samples per tag, which we don't need to do because of our reading abstractions. Moreover, we can just load and filter all data at once to avoid edge artifacts.

It seems like the hits/misses/false alarms are now most strongly tied to how the filtering is done (windowing, steepness, corner frequency) and I think we should just make a sensible choice (probably low-pass below the line freq) and go with that, noting that MF behavior will be different. For now I've pushed something that gets me 53/6/14/7271 for hit/miss/FA/CR on 32 files, with the numbers bouncing around depending mostly on how I change our filtering choices.

@wmvanvliet can you try re-running on your data? You can remove any flat or bad marking steps at the Python end, e.g. you no longer need to `mark_flat` as we now internally mimic what MF does.

File: mne/preprocessing/maxwell.py (hunk 1 above)
Comment (larsoner) [reply mne/preprocessing/maxwell.py]: This is a dev-only thing -- we should only ever have the two options. So I use this kind of like an "active comment" that we know this should be the case. I find it also prevents me from making silly mistakes during development, too.

**Comment by larsoner** [comment]
ping @wmvanvliet :

> can you try re-running on your data? You can remove any flat or bad marking steps at the Python end, e.g. you no longer need to mark_flat as we now internally mimic what MF does.

See my explanation above, but I think at this point even with some mismatches we are doing about as well as we should try to do in matching what MaxFilter does. See if you agree with my reasoning above.

With this and #7290 I'll finally be able to have everything in Python (no need for pushing files to a MaxFilter workstation)...

**Comment by larsoner** [comment]
ping @wmvanvliet, it would be nice to move forward on this if possible

**Comment by wmvanvliet** [comment]
Updates testing script here: https://gist.github.com/wmvanvliet/a3944ad425f28e33fba836315f78169d

Results are here: https://pastebin.com/9uhjC2is

**Comment by wmvanvliet** [comment]
Updated script: https://gist.github.com/wmvanvliet/a3944ad425f28e33fba836315f78169d
Results: https://pastebin.com/Pf53JKVn

**Comment by larsoner** [comment]
@wmvanvliet I think that's pretty good! It's not perfect, but based on what I wrote above I think this is expected. Are you (sufficiently) convinced?

**Comment by wmvanvliet** [comment]
At the very least, it's a useful addition :) It's ok for me if things are not 100% MF compatible, as other portions such as the head position tracking also vary a little.

File: mne/preprocessing/maxwell.py [hunk 3]
```diff
@@ -1853,3 +1887,167 @@ def _trans_sss_basis(exp, all_coils, trans=None, coil_scale=100.):
     S_tot = _sss_basis(exp, all_coils)
     S_tot *= coil_scale
     return S_tot
+
+
+@verbose
+def maxwell_autobad(raw, limit=7., duration=5., min_count=5, **kwargs):
```
Comment (agramfort) [inline mne/preprocessing/maxwell.py]: Can we avoid kwargs in a the public API?

**Comment by larsoner** [comment]
> We also have a def find_outliers(X, threshold=3.0, max_iter=2): in a file called preprocessing/bads.py

This looks like it's supposed to be a private function -- it's undocumented in `python_reference.rst` and only used in `ica` code. I think we should do a deprecation cycle to make it private (very easy by renaming to `_find_outliers` and adding a `find_outliers` with deprecation decocorator).

That frees us up from worrying about that function at all. I'm fine with `find_bads_maxwell` as it does make it so that future `find_bads_*` methods could be added.

**Comment by larsoner** [comment]
Pushed a commit to rename to `find_bad_channels_maxwell`

**Comment by agramfort** [comment]
+1 for find_bad_channels_maxwell and to deprecate find_outliers

**Comment by larsoner** [comment]
CI failures are unrelated

**Comment by agramfort** [comment]
awesome @larsoner !

**Pull request merged**